staging/src/k8s.io/apiserver/plugin/pkg/kmsprovider/vault/vault.go

```diff
@@ -0,0 +1,218 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package vault implements envelop encryption provider based on Vault KMS
+package vault
+
+import (
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"strings"
+	"sync"
+
+	"github.com/ghodss/yaml"
+
+	"k8s.io/apiserver/pkg/storage/value/encrypt/envelope"
+)
+
+const vaultPrefix = "vault"
+
+//EnvelopeConfig contains connection information for Vault transformer
+type EnvelopeConfig struct {
+	// The names of encryption key for Vault transit communication
+	KeyNames []string `json:"key-names"`
+	// Vault listen address, for example https://localhost:8200
+	Address string `json:"addr"`
+
+	// Token authentication information
+	Token string `json:"token"`
+
+	// TLS certificate authentication information
+	ClientCert string `json:"client-cert"`
+	ClientKey  string `json:"client-key"`
+
+	// AppRole authentication information
+	RoleID   string `json:"role-id"`
+	SecretID string `json:"secret-id"`
+
+	// CACert is the path to a PEM-encoded CA cert file to use to verify the
+	// Vault server SSL certificate.
+	CACert string `json:"ca-cert"`
+
+	// TLSServerName, if set, is used to set the SNI host when connecting via TLS.
+	TLSServerName string `json:"tls-server-name"`
+
+	// The path for transit API, default is "transit"
+	TransitPath string `json:"transit-path"`
+
+	// The path for auth backend, default is "auth"
+	AuthPath string `json:"auth-path"`
+}
+
+// KMSFactory function creates Vault KMS service
+func KMSFactory(configFile io.Reader) (envelope.Service, error) {
+	configFileContents, err := ioutil.ReadAll(configFile)
+	if err != nil {
+		return nil, fmt.Errorf("could not read contents: %v", err)
+	}
+
+	var config EnvelopeConfig
+	err = yaml.Unmarshal(configFileContents, &config)
```

Comment:
Please follow the pattern from https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/resourcequota/config.go .  We want to have versioned config, even for our plugins if they are in tree.  It helps when we need to make changes to representation.

Reply:
How should the plugin use version number in the configuration file? What is the general semantics of version number in a configuration file?

Reply:
> How should the plugin use version number in the configuration file? What is the general semantics of version number in a configuration file?

For a config API like this, you'll be supporting it from day one.  I'd use `v1beta1` to start.  It reflects the status of the config file format, not the status of the plugin itself.